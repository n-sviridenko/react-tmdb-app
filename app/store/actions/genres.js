import { createAction } from 'redux-actions';

export const LIST_REQUEST = 'app/genres/LIST_REQUEST';
export const LIST_SUCCESS = 'app/genres/LIST_SUCCESS';
export const LIST_FAILURE = 'app/genres/LIST_FAILURE';

export const listRequest = createAction(LIST_REQUEST);
export const listSuccess = createAction(LIST_SUCCESS);
export const listFailure = createAction(LIST_FAILURE);
