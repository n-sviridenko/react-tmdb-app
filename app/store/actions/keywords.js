import { createAction } from 'redux-actions';

export const SEARCH_REQUEST = 'app/keywords/SEARCH_REQUEST';
export const SEARCH_SUCCESS = 'app/keywords/SEARCH_SUCCESS';
export const SEARCH_FAILURE = 'app/keywords/SEARCH_FAILURE';

export const searchRequest = createAction(SEARCH_REQUEST, null, (query) => ({ key: query }));
export const searchSuccess = createAction(SEARCH_SUCCESS, null, (res, key) => ({ key }));
export const searchFailure = createAction(SEARCH_FAILURE, null, (err, key) => ({ key }));

export const SHOW_REQUEST = 'app/keywords/SHOW_REQUEST';
export const SHOW_SUCCESS = 'app/keywords/SHOW_SUCCESS';
export const SHOW_FAILURE = 'app/keywords/SHOW_FAILURE';

export const showRequest = createAction(SHOW_REQUEST, null, (id) => ({ key: id }));
export const showSuccess = createAction(SHOW_SUCCESS, null, (res, key) => ({ key }));
export const showFailure = createAction(SHOW_FAILURE, null, (err, key) => ({ key }));
