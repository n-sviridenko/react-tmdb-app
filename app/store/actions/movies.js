import { createAction } from 'redux-actions';

export const DISCOVER_REQUEST = 'app/movies/DISCOVER_REQUEST';
export const DISCOVER_SUCCESS = 'app/movies/DISCOVER_SUCCESS';
export const DISCOVER_FAILURE = 'app/movies/DISCOVER_FAILURE';

export const discoverRequest = createAction(DISCOVER_REQUEST);
export const discoverSuccess = createAction(DISCOVER_SUCCESS);
export const discoverFailure = createAction(DISCOVER_FAILURE);
