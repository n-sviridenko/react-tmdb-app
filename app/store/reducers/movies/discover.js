import * as schemas from 'store/schemas';
import paginatedLoaderCreator, {
  createRootSelector,
} from 'store/reducers/creators/paginatedLoader';
import {
  DISCOVER_REQUEST,
  DISCOVER_SUCCESS,
  DISCOVER_FAILURE,
} from 'store/actions/movies';

export default paginatedLoaderCreator({
  types: [
    DISCOVER_REQUEST,
    DISCOVER_SUCCESS,
    DISCOVER_FAILURE,
  ],
});

const getRawRoot = (state) => state.getIn(['movies', 'discover']);
const getRoot = createRootSelector(getRawRoot, schemas.movie);

export {
  getRoot,
};
