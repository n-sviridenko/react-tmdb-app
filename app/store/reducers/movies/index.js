import { combineReducers } from 'redux-immutable';

import discover from './discover';

export default combineReducers({
  discover,
});
