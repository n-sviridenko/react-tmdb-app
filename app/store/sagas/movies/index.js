import { takeLatest } from 'redux-saga/effects';

import * as actions from 'store/actions/movies';
import discover from './discover';

export function* discoverWatcher() {
  yield takeLatest(actions.DISCOVER_REQUEST, discover);
}

export default [
  discoverWatcher,
];
