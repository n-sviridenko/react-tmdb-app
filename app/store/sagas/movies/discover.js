import { call, put } from 'redux-saga/effects';

import * as schemas from 'store/schemas';
import * as actions from 'store/actions/movies';
import { sendListRequest } from 'store/sagas/tmdbClient';

function* discover({ payload: data }) {
  const query = {};

  if (data.get('year')) {
    query.primary_release_year = data.get('year');
  }

  if (data.get('orderBy')) {
    query.sort_by = data.get('orderBy');
  }

  if (data.has('genres') && data.get('genres').size > 0) {
    query.with_genres = data.get('genres').join(',');
  }

  if (data.has('keywords') && data.get('keywords').size > 0) {
    query.with_keywords = data.get('keywords').join(',');
  }

  if (data.get('page')) {
    query.page = data.get('page');
  }

  try {
    const list = yield call(sendListRequest, 'discover.getMovies', schemas.movie, query);

    yield put(actions.discoverSuccess(list));
  } catch (err) {
    yield put(actions.discoverFailure(err));
  }
}

export default discover;
