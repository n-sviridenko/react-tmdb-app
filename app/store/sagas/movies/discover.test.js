/* eslint-disable redux-saga/yield-effects */
import { put } from 'redux-saga/effects';
import { fromJS } from 'immutable';

import * as actions from 'store/actions/movies';
import discover from './discover';

describe('movies discover saga', () => {
  let task;

  beforeEach(() => {
    task = discover({ payload: fromJS({}) });

    const callDescriptor = task.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the discoverSuccess if request was succeeded', () => {
    const list = fromJS({ count: 10, pageCount: 2, page: 1, items: [] });
    const putDescriptor = task.next(list).value;
    expect(putDescriptor).toEqual(put(actions.discoverSuccess(list)));
  });

  it('should dispatch the discoverFailure if request was failed', () => {
    const error = new Error();
    const putDescriptor = task.throw(error).value;
    expect(putDescriptor).toEqual(put(actions.discoverFailure(error)));
  });
});
