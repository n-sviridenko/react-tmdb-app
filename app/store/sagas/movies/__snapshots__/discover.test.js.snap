exports[`movies discover saga should dispatch the discoverFailure if request was failed 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "discover.getMovies",
      EntitySchema {
        "_getId": [Function],
        "_idAttribute": "id",
        "_key": "movie",
        "_mergeStrategy": [Function],
        "_processStrategy": [Function],
        "schema": Object {},
      },
      Object {},
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`movies discover saga should dispatch the discoverSuccess if request was succeeded 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "discover.getMovies",
      EntitySchema {
        "_getId": [Function],
        "_idAttribute": "id",
        "_key": "movie",
        "_mergeStrategy": [Function],
        "_processStrategy": [Function],
        "schema": Object {},
      },
      Object {},
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
